pkg/controller/core/resourceflavor_controller.go

```diff
@@ -18,36 +18,72 @@ package core
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/go-logr/logr"
 	"k8s.io/klog/v2"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 
 	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
 	"sigs.k8s.io/kueue/pkg/cache"
 	"sigs.k8s.io/kueue/pkg/queue"
 )
 
+const ResourceFlavorFinalizerName = "kueue.io/resourceflavor-protection"
+
 // ResourceFlavorReconciler reconciles a ResourceFlavor object
 type ResourceFlavorReconciler struct {
 	log      logr.Logger
 	qManager *queue.Manager
 	cache    *cache.Cache
+	client   client.Client
 }
 
-func NewResourceFlavorReconciler(qMgr *queue.Manager, cache *cache.Cache) *ResourceFlavorReconciler {
+func NewResourceFlavorReconciler(client client.Client, qMgr *queue.Manager, cache *cache.Cache) *ResourceFlavorReconciler {
 	return &ResourceFlavorReconciler{
 		log:      ctrl.Log.WithName("resourceflavor-reconciler"),
 		cache:    cache,
+		client:   client,
 		qManager: qMgr,
 	}
 }
 
-//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch;update;delete
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors/finalizers,verbs=update
 
 func (r *ResourceFlavorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	// Nothing to do here.
+	var flavor kueue.ResourceFlavor
+	if err := r.client.Get(ctx, req.NamespacedName, &flavor); err != nil {
+		// we'll ignore not-found errors, since there is nothing to do.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log := ctrl.LoggerFrom(ctx).WithValues("resourceFlavor", klog.KObj(&flavor))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling ResourceFlavor")
+
+	if flavor.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&flavor, ResourceFlavorFinalizerName) {
+			controllerutil.AddFinalizer(&flavor, ResourceFlavorFinalizerName)
+			if err := r.client.Update(ctx, &flavor); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(&flavor, ResourceFlavorFinalizerName) {
+			if cqName, ok := r.cache.FlavorInUse(flavor.Name); ok {
+				return ctrl.Result{}, fmt.Errorf("flavor %s is in use by clusterQueue %s", flavor.Name, cqName)
+			}
+
+			controllerutil.RemoveFinalizer(&flavor, ResourceFlavorFinalizerName)
+			if err := r.client.Update(ctx, &flavor); err != nil {
```

Comment:
Better to use patch to handle finalizers. A bit more reliable.

Reply:
I would prefer to use `update` as more security, if Update failed, we will requeue.